File: LEETCODE/Hard/Trie/ImplementTrieWithCaching.java

import java.util.HashMap;


class ImplementTrieWithCaching {

    // Implemented using HashMap for child existence checks 
    private class TrieNode {
        HashMap<Character, TrieNode> children;
        Boolean isWordEnd;

        public TrieNode() {
            children = new  HashMap<>();
            isWordEnd = false;
        }
    }

    private TrieNode root;

    public ImplementTrieWithCaching() {
        root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode curr = root;
        for (char c : word.toCharArray()) {
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }

        curr.isWordEnd = true;
    }

    public boolean search(String word) {
        TrieNode curr = root;
        for (char c : word.toCharArray()) {
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        return curr.isWordEnd;
    }

    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for (char c : prefix.toCharArray()) {
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        return true;
    }


    public static void main(String[] args) {
        ImplementTrieWithCaching trie = new ImplementTrieWithCaching();
        trie.insert("apple");
        System.out.println(trie.search("apple"));   // returns true
        System.out.println(trie.search("app"));     // returns false
        System.out.println(trie.startsWith("app")); // returns true
        trie.insert("app");
        System.out.println(trie.search("app"));     // returns true
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: LEETCODE/Hard/Trie/ImplementTrieWithCaching.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <string>
#include <vector>
#include <functional>
#include <unordered_map>
#include <fstream>
using namespace std;


class ImplementTrieWithCaching {
public:
    struct TrieNode {
        char val;
        unordered_map<char, TrieNode*> children;
        bool isEnd;
        TrieNode(char ch) {
            val = ch;
            isEnd = false;
        }
    };

    TrieNode* root;

    ImplementTrieWithCaching() {
        auto init = atexit([]() { ofstream("display_runtime.txt") << "0"; });
		ios_base::sync_with_stdio(0);
		cin.tie(0);
		cout.tie(0);
        root = new TrieNode('\0');
    }

    void insertUtil(string str, TrieNode* node) {
        if(!str.length()) {
            node->isEnd = true;
            return;
        }
        if(node->children[str[0]]) {
            insertUtil(str.substr(1), node->children[str[0]]);
        } else {
            node->children[str[0]] = new TrieNode(str[0]);
            insertUtil(str.substr(1), node->children[str[0]]);
        }
    }
    
    void insert(string word) {
        insertUtil(word, root);
    }

    bool searchUtil(string str, TrieNode* node) {
        if(!str.length()) {
            return node->isEnd ? 1 : 0;
        }
        if(node->children[str[0]]) {
            return searchUtil(str.substr(1), node->children[str[0]]);
        }
        return 0;
    }
    
    bool search(string word) {
        return searchUtil(word, root);
    }

    bool startsWithUtil(string str, TrieNode* node) {
        if(!str.length()) {
            return 1;
        }
        if(node->children[str[0]]) {
            return startsWithUtil(str.substr(1), node->children[str[0]]);
        }
        return 0;
    }
    
    bool startsWith(string prefix) {
        return startsWithUtil(prefix, root);
    }


    void display(TrieNode* node, string curr) {
        if(node->isEnd) {
            cout << curr << "\n";
        }
        for(auto it : node->children) {
            display(it.second, curr + it.first);
        }
    }

    
};
 int main(int argc, char const *argv[])
    {
        ImplementTrieWithCaching trie;
        trie.insert("apple");
        cout << trie.search("apple") << "\n";   // return True
        cout << trie.search("app") << "\n";     // return False
        cout << trie.startsWith("app") << "\n"; // return True
        trie.insert("app");
        cout << trie.search("app") << "\n";     // return True
        trie.display(trie.root, "");
        return 0;
    }


/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */